src/internal/connector/onedrive/restore.go

```diff
@@ -118,50 +146,130 @@ func RestoreCollection(
 		select {
 		case <-ctx.Done():
 			errUpdater("context canceled", ctx.Err())
-			return metrics, true
+			return metrics, folderPerms, true
 
 		case itemData, ok := <-items:
 			if !ok {
-				return metrics, false
-			}
-			metrics.Objects++
-
-			metrics.TotalBytes += int64(len(copyBuffer))
-
-			itemInfo, err := restoreItem(ctx,
-				service,
-				itemData,
-				drivePath.DriveID,
-				restoreFolderID,
-				copyBuffer,
-				source)
-			if err != nil {
-				errUpdater(itemData.UUID(), err)
-				continue
+				return metrics, folderPerms, false
 			}
 
 			itemPath, err := dc.FullPath().Append(itemData.UUID(), true)
 			if err != nil {
 				logger.Ctx(ctx).DPanicw("transforming item to full path", "error", err)
 				errUpdater(itemData.UUID(), err)
-
 				continue
 			}
 
-			deets.Add(
-				itemPath.String(),
-				itemPath.ShortRef(),
-				"",
-				true,
-				itemInfo)
+			if source == OneDriveSource {
+				name := itemData.UUID()
+				if strings.HasSuffix(name, ".data") {
+					metrics.Objects++
+					metrics.TotalBytes += int64(len(copyBuffer))
+					trimmedName := name[:len(name)-5]
+
+					itemID, itemInfo, err = restoreData(ctx, service, trimmedName, itemData, drivePath.DriveID, restoreFolderID, copyBuffer, source)
+					if err != nil {
+						errUpdater(itemData.UUID(), err)
+						continue
+					}
+
+					restoredIDs[trimmedName] = itemID
+
+					deets.Add(itemPath.String(), itemPath.ShortRef(), "", true, itemInfo)
+				} else if strings.HasSuffix(name, ".meta") {
+					meta, err := getMetadata(itemData.ToReader())
+					if err != nil {
+						errUpdater(itemData.UUID(), err)
+						continue
+					}
+
+					trimmedName := name[:len(name)-5]
+					restoreID, ok := restoredIDs[trimmedName]
+					if !ok {
+						errUpdater(itemData.UUID(), fmt.Errorf("item not available to restore permissions"))
+						continue
+					}
+
+					restorePermissions(ctx, service, drivePath.DriveID, restoreID, parentPerms, meta.Permissions)
+
+					// TODO(meain): Finalize on how we count success
+					metrics.Successes++ // Counted as success when we have also restored metadata
+				} else if strings.HasSuffix(name, ".dirmeta") {
+					meta, err := getMetadata(itemData.ToReader())
+					if err != nil {
+						errUpdater(itemData.UUID(), err)
+						continue
+					}
+
+					trimmedName := name[:len(name)-8]
+					createRestoreFolder(ctx, service, drivePath.DriveID, trimmedName, restoreFolderID, parentPerms, meta.Permissions)
+
+					path := itemPath.String()
+					folderPerms[path[:len(path)-8]] = meta.Permissions
+				} else {
+					if !ok {
+						// TODO(meain): support older backups?
+						errUpdater(itemData.UUID(), fmt.Errorf("invalid backup format, you might be using an old backup"))
```

Comment:
Should we consider adding some kind of version information in the backup?

Reply:
Also, do we have to support older backups. There is currently no way of reliably checking if it was backed up using an older version. We could probably change this by changing the format of the details file to include some kind of version information.

Reply:
>Should we consider adding some kind of version information in the backup?

which information are you trying to version? Backups do currently include an "on-disk layout version number" (this was that random leading `0001` you were seeing when fetching folder permissions in one of the other WIP PRs for this feature). That is injected, checked, and stripped off by kopia wrapper during backup/restore. As far as versioning the data stored in the files to tie it to changes in some service code (e.x. OneDrive), we don't really have anything like that yet

I think whether we need something like that is dependent on two things:
* if there is reasonable "fallback" behavior we can implement for when we have no metadata files
* if the fallback behavior makes sense to users of corso

Reply:
I might have to look a bit more into what the "on-disk layout version number" is used for. The version number I am suggesting here would be added to the details file. In this current scenario, it would be used to determine if the files were stored with a `.data` and `.meta` suffix or just a 1:1 map of files. Essentially, some way to determine that the backup was from an older version of corso when we hit this if block.

Reply:
>might have to look a bit more into what the "on-disk layout version number" is used for.

Right now it's not really used for anything. The idea behind it though is if we start doing things like adding data to the front of items (e.x. store OneDrive filename inline with file data or something) then knowing which version of the data you're reading would become important

Reply:
As of now, I'm skipping adding any version. One thing to keep in mind is that once merged corso will fail when trying to restore with an older backup.